src/components/JsonSchemaViewer.tsx

```diff
@@ -1,127 +1,77 @@
-import type { SchemaTreeRefDereferenceFn } from '@stoplight/json-schema-tree';
-import { isRegularNode } from '@stoplight/json-schema-tree';
-import { Box, Flex } from '@stoplight/mosaic';
-import { ErrorBoundaryForwardedProps, FallbackComponent, withErrorBoundary } from '@stoplight/react-error-boundary';
-import { Tree, TreeState, TreeStore } from '@stoplight/tree-list';
+import { isRegularNode, SchemaTree as JsonSchemaTree, SchemaTreeRefDereferenceFn } from '@stoplight/json-schema-tree';
+import { Box, VStack } from '@stoplight/mosaic';
+import { ErrorBoundaryForwardedProps, FallbackProps, withErrorBoundary } from '@stoplight/react-error-boundary';
 import cn from 'classnames';
 import type { JSONSchema4 } from 'json-schema';
-import { action } from 'mobx';
 import * as React from 'react';
 
-import { SchemaTreeContext, ViewModeContext } from '../contexts';
-import { SchemaTreeListTree, SchemaTreeOptions } from '../tree';
-import type { GoToRefHandler, RowRenderer, ViewMode } from '../types';
-import { SchemaTree as SchemaTreeComponent } from './SchemaTree';
+import { JSVOptions, JSVOptionsContextProvider } from '../contexts';
+import { SchemaRow } from './SchemaRow';
 
-export interface IJsonSchemaViewer {
+export type JsonSchemaProps = Partial<JSVOptions> & {
   schema: JSONSchema4;
   emptyText?: string;
-  defaultExpandedDepth?: number;
   className?: string;
-  maxRows?: number;
-  onGoToRef?: GoToRefHandler;
-  mergeAllOf?: boolean;
-  FallbackComponent?: typeof FallbackComponent;
-  rowRenderer?: RowRenderer;
   resolveRef?: SchemaTreeRefDereferenceFn;
-  viewMode?: ViewMode;
-}
-
-export class JsonSchemaViewerComponent extends React.PureComponent<
-  IJsonSchemaViewer & ErrorBoundaryForwardedProps,
-  { isEmpty: boolean }
-> {
-  protected readonly treeStore: TreeStore;
-  protected readonly tree: SchemaTreeListTree;
-  protected readonly treeState: TreeState;
-
-  public readonly state = {
-    isEmpty: true,
-  };
-
-  constructor(props: IJsonSchemaViewer & ErrorBoundaryForwardedProps) {
-    super(props);
+};
 
-    this.treeState = new TreeState();
-    this.tree = new SchemaTreeListTree(props.schema, this.treeState, this.treeOptions);
-    this.treeStore = new TreeStore(this.tree, this.treeState, {
-      defaultExpandedDepth: this.expandedDepth,
+const JsonSchemaViewerComponent: React.FC<JsonSchemaProps & ErrorBoundaryForwardedProps> = ({
+  schema,
+  viewMode = 'standalone',
+  className,
+  resolveRef,
+  emptyText = 'No schema defined',
+  defaultExpandedDepth = 2,
+  onGoToRef,
+  renderRowAddon,
+}) => {
+  const jsonSchemaTreeRoot = React.useMemo(() => {
+    const jsonSchemaTree = new JsonSchemaTree(schema, {
+      mergeAllOf: true,
+      refResolver: resolveRef,
     });
-  }
+    jsonSchemaTree.walker.hookInto('filter', node => {
```

Comment:
You lost `stepIn` in the transition.
This is quite important, otherwise _really_ large schemas with a ton of $refs will be much slower to process than needed.
`stepIn` is needed, so that JST tree is not built upfront for the whole schema, but for actual parts that are currently rendered.
This is also one of the more notable factors contributing to the entire complexity of the former solution.
While, getting rid of virtualization is fine by me, we certainly should at least try to preserve that optimization.
By adding it back, we should be more or less on par with current v3.